### PR TITLE
feat(de form): add field rules

### PR DESCRIPTION
**Pull request by Mohammer5**
This is a draft PR as I want to discuss moving validation from the form-level to the field-level (see last paragraph below).

- [x] Implements the field rules for the data element form
- [x] Tests
- [x] Docs

The old app uses a configuration to define the rules. I tried to follow a similar approach but aligned with how React works and realized that it adds a lot of complexity:

- The Zod schema has to be created on the view-component level.
- The category combo is optional when `domainType === 'TRACKER'`.
- The disabled state would have to live in the views.
- The Zod schema would have to be dynamic (category combo optional / required).
- The disabled state would have to be passed down to the actual category combo field (either directly, with context, or React-external state like Zustand).

Instead, I've co-located everything with the actual field, which makes it somewhat simple because everything related to a particular field is in one place only. It doesn't require too much state (e.g., whether the category combo is disabled or not is simply `values.domainType === 'TRACKER'`). This approach won't yield a list of rules, which requires us to document which rules exist. I've done that in comments above the fields that have rules in `fields.tsx`.

I'm tempted to remove Zod and move validation to the field level. This will further untangle/simplify the architecture as field-level validator functions will be co-located with the fields in the code. The biggest advantage that Zod provides is that we don't have to duplicate types, but this doesn't apply to us as we already have generated types. Having the validators on the field level allows us to have very simple validation functions where necessary (see the new `validateCategoryCombo` function in `fields.tsx`) which reduces cognitive overhead. With the current approach (with Zod), you'll have to remember the schema itself, where the schema is located (`dateElementsSchema.ts`), how it's being applied (on RFF's `<Form />` component), and where the form component is located (in the views). This will all be reduced to `it's where the field is` (in `fields.tsx`).

### CODE REVIEW

File: src/pages/dataElements/form/fields.tsx [hunk 1]
```diff
@@ -328,33 +388,127 @@ export function LegendSetField() {
...
+            id: values.optionSet.id,
+            fields: valueTypeOptionSetFields,
+        },
+    })
+
+    useEffect(() => {
+        if (values.optionSet.id) {
+            refetch({ id: values.optionSet.id })
+        }
+    }, [refetch, values.optionSet.id])
```
Comment (Birkbjo) [inline src/pages/dataElements/form/fields.tsx:424]: I'm a bit worried about the need of doing this, but it does make sense - especially since `valueType` is not present in the optionSet request for now - but ideally we could just reuse that and set cacheTime to not send new requests in this field. I'm just worried about the requests that would go out if a lot of fields would need to do something like this. I'm also not a fan of the fetch logic in this field - but that is something we could put in a hook - but not sure if it's worth it here?

File: src/pages/dataElements/form/fields.tsx [hunk 2]
```diff
@@ -373,19 +532,40 @@ export function AggregationTypeField() {
...
+const validateCategoryCombo = (value: { id: string }) => {
+    if (!value) {
+        return i18n.t('Required')
+    }
+}
+
+/*
+ * @TODO: Verify that the api ignores the category combo when it's disabled.
+ *        If it does not, file a jira issue and "escalate" this so it will be
+ *        implemented
```
Comment (Birkbjo) [inline src/pages/dataElements/form/fields.tsx:557]: I think we should just default to the default catcombo?

File: src/pages/dataElements/form/fields.tsx [hunk 3]
```diff
@@ -428,10 +609,19 @@ export function CategoryComboField() {
...
+ */
 export function OptionSetField() {
     const newOptionSetLink = useHref('/optionSets/new')
-    const { input, meta } = useField('optionSet.id', {
+    // Not using a dot path because setting the value to an empty string
+    // removes the option set from the form state entirely
+    const { input, meta } = useField('optionSet', {
         validateFields: [],
+        format: (optionSet) => optionSet.id,
+        parse: (id) => ({ id }),
```
Comment (Birkbjo) [inline src/pages/dataElements/form/fields.tsx:624]: I actually think this can be a good pattern - eg. keep the whole object/model and use format and parse. It could allow us to not have to fetch for dependent rules- and fetch more in one request in either form-load (initialize form) - or in specific fields. It could however make it a bit hard to follow what fields are actually there - so we would have to look at typings for it.

File: src/pages/dataElements/form/fields.tsx (hunk 1 above)
Comment (Mohammer5) [reply src/pages/dataElements/form/fields.tsx:424]: >  but ideally we could just reuse that and set cacheTime to not send new requests in this field

That's what @ismay suggested as well. The problem is that the option sets request from the `OptionSetSelect` is paginated, so we'd have to know which page that particular option set is in and simply requesting what we need here felt a lot simpler.

Another point for doing a request here is that we colocate the data dependency with where it's actually used/needed. This way we don't put a field that we don't need in the option set field into the requests for its options (which might actually result in higher bandwidth usage than the request proposed by this PR).

> I'm also not a fan of the fetch logic in this field - but that is something we could put in a hook - but not sure if it's worth it here?

Yet another case for just using react-query because it supports `enabled` rather than `lazy`, which would remove the fetch logic entirely. But I didn't use it because we're not allowed :upside_down_face: 

> I'm just worried about the requests that would go out if a lot of fields would need to do something like this.

I'm not worried about this until we have to be. These requests should transfer only a couple of bytes and are used by administrators only so that - even when the entire dhis2 instance is used with a limited amount of data - this should be next to insignificant. I def. agree that once we have heavy requests we should think about whether it makes sense to do it the same way as here, but I wouldn't optimize something that might not need optimization after all.

File: src/pages/dataElements/form/fields.tsx (hunk 3 above)
Comment (Mohammer5) [reply src/pages/dataElements/form/fields.tsx:624]: > keep the whole object/model

That's already the case (the validation PR also uses the entire object. The form state is basically a replicate of the model).
I just noticed that when selecting an empty value, the entire object is thrown out of RFF's form state (it caused `values.optionSet.id` to throw as I was attempting to read `id` of `undefined`).

> It could however make it a bit hard to follow what fields are actually there - so we would have to look at typings for it.

I was thinking about this, too. To me this issue disappeared when I realized that I could simply look at the creation of the initial state as that's already the complete state (either with or without values, depending on whether it's the edit or new view)

**Comment by Birkbjo** [comment]
> I'm tempted to remove Zod and move validation to the field level. This will further untangle/simplify the architecture as field-level validator functions will be co-located with the fields in the code. The biggest advantage that Zod provides is that we don't have to duplicate types, but this doesn't apply to us as we already have generated types. Having the validators on the field level allows us to have very simple validation functions where necessary 

I see what you mean, but I do think there can be some value in having zod. The field-level rules could even just refer to those fields - so every validation of a form is co-located. We might also not need field-level validations, if the form-level is run regardless?

File: src/pages/dataElements/form/fields.tsx (hunk 1 above)
Comment (Birkbjo) [reply src/pages/dataElements/form/fields.tsx:429]: > That's what @ismay suggested as well. The problem is that the option sets request from the OptionSetSelect is paginated, so we'd have to know which page that particular option set is in and simply requesting what we need here felt a lot simpler.

I see, that makes sense. 

> Yet another case for just using react-query because it supports enabled rather than lazy, which would remove the fetch logic entirely. 

Agreed, to be pedantic though - it's due to refetch and static queries - lazy is just a "rename" of enabled. But yeah, it's due to how we have to imperatively call refetch. 

> Another point for doing a request here is that we colocate the data dependency with where it's actually used/needed.

Here it's just one. But some other fields (in other sections) will have a lot of dependencies, to the point that I think it would make sense to fetch all of those in one request - even could make sense to do it upfront. But it could potentially look the same - and extract it to a hook that is used across multiple fields. But lets keep this for now.

**Comment by Mohammer5** [comment]
We've decided to remove validation / require-checks from the zod schema and move that to the field level.
Zod is now being used to verify the shape as well as do post-submit transformations like trimming.

File: src/pages/dataElements/fields/NameField.tsx [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+    const params = useParams()
+    const dataElementId = params.id as string
+    const checkIsValueTaken = useIsFieldValueUnique({
+        model: 'dataElements',
+        field: 'name',
+        id: dataElementId,
+    })
+
+    const checkMaxLength = useCheckMaxLengthFromSchema(
+        'dataElement' as SchemaName,
```
Comment (Birkbjo) [inline src/pages/dataElements/fields/NameField.tsx:28]: Change import to `import { SchemaName} `, and use `SchemaName.dataElement` instead?

File: src/pages/dataElements/fields/NameField.tsx [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+        'name'
+    )
+
+    return useMemo(
+        () =>
+            composeAsyncValidators<string, FormValues>([
+                checkIsValueTaken as NameValidator,
```
Comment (Birkbjo) [inline src/pages/dataElements/fields/NameField.tsx:35]: I don't think this type cast is needed?

File: src/lib/form/composeAsyncValidators/types.ts [hunk 6]
```diff
@@ -0,0 +1,6 @@
+type ReturnValue = string | undefined
+
+export type Validator<Value, FormValues = object> = (
```
Comment (Birkbjo) [inline src/lib/form/composeAsyncValidators/types.ts:3]: I think it would be more clear if this was name something like `FormFieldValidator`. We already have a type from `ui` called `Validator`.

File: src/lib/models/useIsFieldValueUnique.ts [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+
+                return { pageSize: 1, fields: 'id', filter }
+            },
+        },
+    })
+    const engine = useDataEngine()
+
+    const memoized = useMemo(
+        () =>
+            memoize(async (value: string) => {
```
Comment (Birkbjo) [inline src/lib/models/useIsFieldValueUnique.ts:41]:
```suggestion
            memoize(async (value?: string) => {
```

**Comment by Mohammer5** [comment]
@Birkbjo nice finds! I've implement all change requests!

**Review by Birkbjo — APPROVED** [review]
